File: day3/part1.py
with open("part1.txt") as f:
    f.readline()
    line = f.readline()
    width = len(line) - 1
    dist = 3  # horizontal distance
    trees = 0
    while line != '':
        if line[dist] == '#':
            trees += 1
        dist = (dist + 3) % width
        line = f.readline()
    print(trees)

File: day4/part1.ts
import * as readline from 'readline'
import * as fs from 'fs'

const reader = readline.createInterface(fs.createReadStream('part1.txt'))

const expectedFields = new Set([
  'byr','iyr','eyr','hgt','hcl','ecl','pid',
])

async function process() {
  let expectedFieldsCount = 0
  let valid = 0
  for await (const line of reader) {
    if(line != '') {
      line.split(' ').forEach((kv) => {
        const key = kv.split(':')[0]
        if(expectedFields.has(key)) expectedFieldsCount++
      })
    } else {
      if(expectedFieldsCount == expectedFields.size) valid++
      expectedFieldsCount = 0
    }
  }
  if(expectedFieldsCount == expectedFields.size) valid++
  console.log(valid)
}

process()